In [536]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.graph_objects as go
import plotly.express as px

Reading the Data : 

In [537]:
data_path = "chicago-traffic-dataset"
traffic_crashes_people = pd.read_csv(data_path+"/traffic-crashes-people.csv",low_memory=False)
df_vehicles = pd.read_csv(data_path+"/traffic-crashes-vehicles.csv",low_memory=False)

Let's first getting fimilar with our dataset .

In [538]:
traffic_crashes_people.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1808605 entries, 0 to 1808604
Data columns (total 29 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   PERSON_ID              object 
 1   PERSON_TYPE            object 
 2   CRASH_RECORD_ID        object 
 3   VEHICLE_ID             float64
 4   CRASH_DATE             object 
 5   SEAT_NO                float64
 6   CITY                   object 
 7   STATE                  object 
 8   ZIPCODE                object 
 9   SEX                    object 
 10  AGE                    float64
 11  DRIVERS_LICENSE_STATE  object 
 12  DRIVERS_LICENSE_CLASS  object 
 13  SAFETY_EQUIPMENT       object 
 14  AIRBAG_DEPLOYED        object 
 15  EJECTION               object 
 16  INJURY_CLASSIFICATION  object 
 17  HOSPITAL               object 
 18  EMS_AGENCY             object 
 19  EMS_RUN_NO             object 
 20  DRIVER_ACTION          object 
 21  DRIVER_VISION          object 
 22  PHYSICAL_CONDITION

In [539]:
traffic_crashes_people.head()

PERSON_ID PERSON_TYPE                                    CRASH_RECORD_ID  \
0   O749947      DRIVER  81dc0de2ed92aa62baccab641fa377be7feb1cc47e6554...   
1   O871921      DRIVER  af84fb5c8d996fcd3aefd36593c3a02e6e7509eeb27568...   
2    O10018      DRIVER  71162af7bf22799b776547132ebf134b5b438dcf3dac6b...   
3    O10038      DRIVER  c21c476e2ccc41af550b5d858d22aaac4ffc88745a1700...   
4    O10039      DRIVER  eb390a4c8e114c69488f5fb8a097fe629f5a92fd528cf4...   

   VEHICLE_ID              CRASH_DATE  SEAT_NO     CITY STATE ZIPCODE SEX  \
0    834816.0  09/28/2019 03:30:00 AM      NaN  CHICAGO    IL   60651   M   
1    827212.0  04/13/2020 10:50:00 PM      NaN  CHICAGO    IL   60620   M   
2      9579.0  11/01/2015 05:00:00 AM      NaN      NaN   NaN     NaN   X   
3      9598.0  11/01/2015 08:00:00 AM      NaN      NaN   NaN     NaN   X   
4      9600.0  11/01/2015 10:15:00 AM      NaN      NaN   NaN     NaN   X   

    AGE DRIVERS_LICENSE_STATE DRIVERS_LICENSE_CLASS  SAFETY_EQUIPMENT  \
0  25.0                    IL                     D      NONE PRESENT   
1  37.0                    IL                   NaN  SAFETY BELT USED   
2   NaN                   NaN                   NaN     USAGE UNKNOWN   
3   NaN                   NaN                   NaN     USAGE UNKNOWN   
4   NaN                   NaN                   NaN     USAGE UNKNOWN   

      AIRBAG_DEPLOYED EJECTION    INJURY_CLASSIFICATION HOSPITAL EMS_AGENCY  \
0  DEPLOYMENT UNKNOWN     NONE  NO INDICATION OF INJURY      NaN        NaN   
1      DID NOT DEPLOY     NONE  NO INDICATION OF INJURY      NaN        NaN   
2  DEPLOYMENT UNKNOWN     NONE  NO INDICATION OF INJURY      NaN        NaN   
3  DEPLOYMENT UNKNOWN  UNKNOWN  NO INDICATION OF INJURY      NaN        NaN   
4  DEPLOYMENT UNKNOWN  UNKNOWN  NO INDICATION OF INJURY      NaN        NaN   

  EMS_RUN_NO     DRIVER_ACTION DRIVER_VISION PHYSICAL_CONDITION  \
0        NaN           UNKNOWN       UNKNOWN            UNKNOWN   
1        NaN              NONE  NOT OBSCURED             NORMAL   
2        NaN  IMPROPER BACKING       UNKNOWN            UNKNOWN   
3        NaN           UNKNOWN       UNKNOWN            UNKNOWN   
4        NaN           UNKNOWN       UNKNOWN            UNKNOWN   

  PEDPEDAL_ACTION PEDPEDAL_VISIBILITY PEDPEDAL_LOCATION        BAC_RESULT  \
0             NaN                 NaN               NaN  TEST NOT OFFERED   
1             NaN                 NaN               NaN  TEST NOT OFFERED   
2             NaN                 NaN               NaN  TEST NOT OFFERED   
3             NaN                 NaN               NaN  TEST NOT OFFERED   
4             NaN                 NaN               NaN  TEST NOT OFFERED   

   BAC_RESULT VALUE CELL_PHONE_USE  
0               NaN            NaN  
1               NaN            NaN  
2               NaN            NaN  
3               NaN            NaN  
4               NaN            NaN

It's a huge dataset with 1808605 and 29 columns , so first we would change the defualt display of tables to show all columns and more rows.


In [540]:
pd.set_option("display.max_columns", None)
pd.set_option('display.max_rows', 100)

let's see where my all datasets meeet with each other, so what the common columns between them ?

In [541]:
common_columns = set(traffic_crashes_people.columns).intersection(df_vehicles.columns)
if common_columns:
    print(f"Shared columns: {', '.join(common_columns)}")
else:
    print("No shared columns found.")

Shared columns: CRASH_RECORD_ID, CRASH_DATE, VEHICLE_ID


In [542]:
common_columns = set(traffic_crashes_people.columns).intersection(df_crashes.columns)
if common_columns:
    print(f"Shared columns: {', '.join(common_columns)}")
else:
    print("No shared columns found.")

NameError: name 'df_crashes' is not defined

let's study the data step by step and dealing with missing values whilethrough. so let's see the count of missing values and it's proportion in each column . 

In [543]:
info_df = pd.DataFrame()
info_df['missing_val'] = traffic_crashes_people.isnull().sum()
info_df['missing_val_ratio'] = (info_df['missing_val'] / traffic_crashes_people.shape[0] * 100).round().astype(int)
info_df

missing_val  missing_val_ratio
PERSON_ID                        0                  0
PERSON_TYPE                      0                  0
CRASH_RECORD_ID                  0                  0
VEHICLE_ID                   36177                  2
CRASH_DATE                       0                  0
SEAT_NO                    1443028                 80
CITY                        492062                 27
STATE                       472632                 26
ZIPCODE                     600317                 33
SEX                          29978                  2
AGE                         527939                 29
DRIVERS_LICENSE_STATE       748686                 41
DRIVERS_LICENSE_CLASS       922585                 51
SAFETY_EQUIPMENT              5061                  0
AIRBAG_DEPLOYED              34799                  2
EJECTION                     22453                  1
INJURY_CLASSIFICATION          729                  0
HOSPITAL                   1509447                 83
EMS_AGENCY                 1623522                 90
EMS_RUN_NO                 1777884                 98
DRIVER_ACTION               368819                 20
DRIVER_VISION               369365                 20
PHYSICAL_CONDITION          367836                 20
PEDPEDAL_ACTION            1774170                 98
PEDPEDAL_VISIBILITY        1774235                 98
PEDPEDAL_LOCATION          1774172                 98
BAC_RESULT                  367837                 20
BAC_RESULT VALUE           1806583                100
CELL_PHONE_USE             1807446                100

#### We have huge loss in our dataset 
let's treat that step by step. and we start with VEHCILE_ID column with has only 2% loss in the whole dataset. 

In [544]:
traffic_crashes_people['PERSON_TYPE'].unique()

array(['DRIVER', 'PASSENGER', 'PEDESTRIAN', 'BICYCLE',
       'NON-MOTOR VEHICLE', 'NON-CONTACT VEHICLE'], dtype=object)

In [545]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == "PEDESTRIAN")]['VEHICLE_ID'].isnull().sum()

21142

In [546]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == "PEDESTRIAN")].shape[0]

21142

In [547]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == "BICYCLE")]['VEHICLE_ID'].isnull().sum()

12725

In [548]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == "BICYCLE")].shape[0]

12725

In [549]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == "NON-MOTOR VEHICLE")]['VEHICLE_ID'].isnull().sum()

1474

In [550]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == "NON-MOTOR VEHICLE")].shape[0]

1474

All Pedestrians and Bicyclists and NON_Motor dosen't have a VEHICLE_ID value. 
of course we will not drop them cause if we do we loose all the people with this type and loose special categories. 
so we can give them a random unqiue value for each row , but that not right cause these values not carrying a vehicle_id in real life , in addition to if we gave them sequential values in this dataset we have to find them in the other datasets like df_vehicles and give them the same ID value we give them here.
So, we can give them all a non_meaning value for IDs like (-1)

In [551]:
traffic_crashes_people.loc[traffic_crashes_people['PERSON_TYPE'] == "BICYCLE", 'VEHICLE_ID'] = -1.0
traffic_crashes_people.loc[traffic_crashes_people['PERSON_TYPE'] == "PEDESTRIAN", 'VEHICLE_ID'] = -1.0
traffic_crashes_people.loc[traffic_crashes_people['PERSON_TYPE'] == "NON-MOTOR VEHICLE", 'VEHICLE_ID'] = -1.0


In [552]:
traffic_crashes_people['VEHICLE_ID'].isnull().sum()

836

We still have missing values for VEHICLE_ID , let's see where. 

In [553]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == "DRIVER")].shape[0]

1407381

I have 699 nan values of VEHICLE_ID for DRIVERS , let's see the VEHICLE_ID values when the person being a PASSENGER

In [554]:
traffic_crashes_people[traffic_crashes_people['PERSON_TYPE'] == "PASSENGER"]['VEHICLE_ID'].isnull().sum()

0

As we expected no nan values in VEHICLE_ID when the person is PASSENGER , so let's try to extract some missed values for VEHICLE_ID for the DRIVERS out of PASSENGERS VEHICLE_ID values when existing in same CRASH_RECORD_ID , we about to see if that is possible.

In [555]:
unique_passenger_Vids = traffic_crashes_people[traffic_crashes_people['PERSON_TYPE'] == "PASSENGER"]['VEHICLE_ID'].unique()
len(unique_passenger_Vids)

248592

In [556]:
unique_driver_Vids = traffic_crashes_people[traffic_crashes_people['PERSON_TYPE'] == "DRIVER"]['VEHICLE_ID'].unique()
len(unique_driver_Vids)

1406683

We have now two series for unique VEHICLE_ID in PASSENGERS and in DRIVERS, so first we will see if there any PASSENGER VEHICLE_ID exist in PASSENGERS series and not in DRIVERS series.

In [557]:
passenger_set = set(unique_passenger_Vids)
driver_set = set(unique_driver_Vids)

if passenger_set.issubset(driver_set):
    print('All values in unique_passenger_Vids are present in unique_driver_Vids')
else:
    print('Not all values in unique_passenger_Vids are present in unique_driver_Vids')

Not all values in unique_passenger_Vids are present in unique_driver_Vids


It seems there are some values not shared between two sets , so there are some missed VEHICLE_ID for DRIVERS maybe we can assign the PASSENGER's VEHICLE_ID value for the DRIVER's nan VEHICLE_ID value in same CRASH_RECORD_ID (but of course when we have no another DRIVER carries this value cause we may have multiple DRIVERS in the same CRASH_RECORD_ID so we can't know the missed DRIVER's VEHICLE_ID for the other drivers or when we have multiple DRIVERS in the same accident so we can't know this PASSENGER share the same vehicle with a sepcific DRIVER).

In [558]:
missing_passenger_Vids = passenger_set - driver_set

if missing_passenger_Vids:
#     print(len(missing_passenger_Vids))
    print('The following VEHICLE_IDs are unique to unique_passenger_Vids:', missing_passenger_Vids)
else:
    print('All values in unique_passenger_Vids are present in unique_driver_Vids')

The following VEHICLE_IDs are unique to unique_passenger_Vids: {458752.0, 1212420.0, 1114117.0, 1638406.0, 1245190.0, 327686.0, 655369.0, 917509.0, 1703947.0, 786443.0, 917515.0, 589841.0, 1441809.0, 1441811.0, 851990.0, 1277978.0, 557084.0, 393244.0, 1277982.0, 1114146.0, 1114147.0, 1310756.0, 852005.0, 852006.0, 426021.0, 294954.0, 983083.0, 1671213.0, 229427.0, 1572916.0, 1343544.0, 1376314.0, 1114171.0, 98363.0, 327742.0, 1409092.0, 917572.0, 131143.0, 1671241.0, 557130.0, 229451.0, 32844.0, 393298.0, 83.0, 32851.0, 753754.0, 327776.0, 1540203.0, 589932.0, 1114226.0, 295028.0, 557176.0, 1671289.0, 1409145.0, 1212539.0, 1212540.0, 1704061.0, 196734.0, 917631.0, 589954.0, 1114248.0, 1147017.0, 1245323.0, 1671308.0, 1212556.0, 1147022.0, 557199.0, 163980.0, 721037.0, 1245328.0, 983179.0, 295060.0, 393364.0, 131222.0, 1563123.0, 786585.0, 721050.0, 229530.0, 884892.0, 1474715.0, 1638558.0, 1474720.0, 884898.0, 32935.0, 753834.0, 950447.0, 884914.0, 1704115.0, 524469.0, 1343673.0, 14092

In [559]:
traffic_crashes_people[( traffic_crashes_people['VEHICLE_ID'].isin(missing_passenger_Vids) ) & (traffic_crashes_people['PERSON_TYPE'] == "DRIVER")]

Empty DataFrame
Columns: [PERSON_ID, PERSON_TYPE, CRASH_RECORD_ID, VEHICLE_ID, CRASH_DATE, SEAT_NO, CITY, STATE, ZIPCODE, SEX, AGE, DRIVERS_LICENSE_STATE, DRIVERS_LICENSE_CLASS, SAFETY_EQUIPMENT, AIRBAG_DEPLOYED, EJECTION, INJURY_CLASSIFICATION, HOSPITAL, EMS_AGENCY, EMS_RUN_NO, DRIVER_ACTION, DRIVER_VISION, PHYSICAL_CONDITION, PEDPEDAL_ACTION, PEDPEDAL_VISIBILITY, PEDPEDAL_LOCATION, BAC_RESULT, BAC_RESULT VALUE, CELL_PHONE_USE]
Index: []

let's check for a random value in missing_passenger_Vids 

In [560]:
traffic_crashes_people[traffic_crashes_people['VEHICLE_ID'] == 1212420.0]


PERSON_ID PERSON_TYPE  \
1271734   P284715   PASSENGER   
1271735   P284716   PASSENGER   

                                           CRASH_RECORD_ID  VEHICLE_ID  \
1271734  f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0...   1212420.0   
1271735  f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0...   1212420.0   

                     CRASH_DATE  SEAT_NO           CITY STATE ZIPCODE SEX  \
1271734  02/04/2022 02:35:00 PM      1.0  DOWNERS GROVE    IL   60515   F   
1271735  02/04/2022 02:35:00 PM     10.0  DOWNERS GROVE    IL   60515   F   

          AGE DRIVERS_LICENSE_STATE DRIVERS_LICENSE_CLASS  SAFETY_EQUIPMENT  \
1271734  69.0                   NaN                   NaN  SAFETY BELT USED   
1271735   NaN                   NaN                   NaN     USAGE UNKNOWN   

        AIRBAG_DEPLOYED EJECTION    INJURY_CLASSIFICATION HOSPITAL EMS_AGENCY  \
1271734  NOT APPLICABLE     NONE  NO INDICATION OF INJURY      NaN        NaN   
1271735  NOT APPLICABLE     NONE  NO INDICATION OF INJURY      NaN        NaN   

        EMS_RUN_NO DRIVER_ACTION DRIVER_VISION PHYSICAL_CONDITION  \
1271734        NaN           NaN           NaN                NaN   
1271735        NaN           NaN           NaN                NaN   

        PEDPEDAL_ACTION PEDPEDAL_VISIBILITY PEDPEDAL_LOCATION BAC_RESULT  \
1271734             NaN                 NaN               NaN        NaN   
1271735             NaN                 NaN               NaN        NaN   

         BAC_RESULT VALUE CELL_PHONE_USE  
1271734               NaN            NaN  
1271735               NaN            NaN

Wait! no DRIVER here 

In [561]:
traffic_crashes_people[traffic_crashes_people['VEHICLE_ID'] == 1212420.0]['CRASH_RECORD_ID']


1271734    f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0...
1271735    f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0...
Name: CRASH_RECORD_ID, dtype: object

In [562]:
traffic_crashes_people[traffic_crashes_people['VEHICLE_ID'] == 1212420.0]['CRASH_RECORD_ID'][1271734]


'f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0b377763f268a3c3c3cce0ec8e82dc0d33242a168b11fd1bf8fa06582807f5e4b727fb8a1be0be66458'

In [563]:
traffic_crashes_people[traffic_crashes_people['CRASH_RECORD_ID'] == 'f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0b377763f268a3c3c3cce0ec8e82dc0d33242a168b11fd1bf8fa06582807f5e4b727fb8a1be0be66458']

PERSON_ID PERSON_TYPE  \
494870   O1277023      DRIVER   
1271734   P284715   PASSENGER   
1271735   P284716   PASSENGER   

                                           CRASH_RECORD_ID  VEHICLE_ID  \
494870   f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0...   1212419.0   
1271734  f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0...   1212420.0   
1271735  f2e73538e2b985ccadccbd66f29c4be8993c9c07868ff0...   1212420.0   

                     CRASH_DATE  SEAT_NO           CITY STATE ZIPCODE SEX  \
494870   02/04/2022 02:35:00 PM      NaN        CHICAGO    IL   60620   F   
1271734  02/04/2022 02:35:00 PM      1.0  DOWNERS GROVE    IL   60515   F   
1271735  02/04/2022 02:35:00 PM     10.0  DOWNERS GROVE    IL   60515   F   

          AGE DRIVERS_LICENSE_STATE DRIVERS_LICENSE_CLASS  SAFETY_EQUIPMENT  \
494870   28.0                   NaN                   NaN  SAFETY BELT USED   
1271734  69.0                   NaN                   NaN  SAFETY BELT USED   
1271735   NaN                   NaN                   NaN     USAGE UNKNOWN   

            AIRBAG_DEPLOYED EJECTION    INJURY_CLASSIFICATION HOSPITAL  \
494870   DEPLOYMENT UNKNOWN     NONE  NO INDICATION OF INJURY      NaN   
1271734      NOT APPLICABLE     NONE  NO INDICATION OF INJURY      NaN   
1271735      NOT APPLICABLE     NONE  NO INDICATION OF INJURY      NaN   

        EMS_AGENCY EMS_RUN_NO DRIVER_ACTION DRIVER_VISION PHYSICAL_CONDITION  \
494870         NaN        NaN         OTHER       UNKNOWN             NORMAL   
1271734        NaN        NaN           NaN           NaN                NaN   
1271735        NaN        NaN           NaN           NaN                NaN   

        PEDPEDAL_ACTION PEDPEDAL_VISIBILITY PEDPEDAL_LOCATION  \
494870              NaN                 NaN               NaN   
1271734             NaN                 NaN               NaN   
1271735             NaN                 NaN               NaN   

               BAC_RESULT  BAC_RESULT VALUE CELL_PHONE_USE  
494870   TEST NOT OFFERED               NaN            NaN  
1271734               NaN               NaN            NaN  
1271735               NaN               NaN            NaN

We found from above that for the same CRASH_RECORD_ID we may have PASSENGERS which belong to a vehicle and another DRIVER that own another VEHICLE completely, and no DRIVER person for the VEHICLE carries PASSENGERS here . 
So we conclude from that it's not neccessary to give the missed VEHICLE_ID value for a DRIVER (even if i have one DRIVER in this accident and the PASSGENR's VEHICLE_IDs are known). so there is not a condition to have a DRIVER when i have PASSENGERS in my data. so we can create a rule which assign the missed values of VEHICLE_ID in DRIVERS from PASSENGERs VEHICLE_ID values.

We can now assign random unique values for the missed VEHICLE_ID in DRIVERS , and assign the same value for the one exist in df_vehicles for the same CRASH_RECORD_ID, but not when i have multiple DRIVERS except if i need to disribute these values randomly on the DRIVERS in df_vehicles for the matched CRASH_RECORD_ID there .
we know that VEHICLE_ID not a feature for any model i will use in future but it is a very sensitive column when it mathces beetween the datasets so changing the reality of the datasets. so let's drop them .

In [564]:
traffic_crashes_people = traffic_crashes_people.dropna(subset=['VEHICLE_ID'])

The next column is SEAT_NO with 80% loss in data. We see that it represents where the person sit in vehicle , So where the missing values ? 

In [565]:
traffic_crashes_people['SEAT_NO'].value_counts()

3.0     176720
6.0      61481
4.0      48684
5.0      17200
1.0      16292
2.0      15337
12.0      9592
7.0       8427
10.0      8175
11.0      3222
8.0        447
Name: SEAT_NO, dtype: int64

In [566]:
traffic_crashes_people[traffic_crashes_people['PERSON_TYPE'] == 'PASSENGER']['SEAT_NO'].isnull().sum()

0

In [567]:
traffic_crashes_people[traffic_crashes_people['PERSON_TYPE'] == "DRIVER"]['SEAT_NO'].isnull().sum()

1406682

In [568]:
traffic_crashes_people[traffic_crashes_people['PERSON_TYPE'] == "DRIVER"]['SEAT_NO'].shape[0]

1406682

the huge miss exists in DRIVER type , all the drivers doesn't have value for SEAT_NO .
but the driver still sit in the vehicle , so we will fill them with the unique value (0) reflect the zero SEAT_NO (_the first one_).

In [569]:
traffic_crashes_people.loc[traffic_crashes_people['PERSON_TYPE'] == "DRIVER", 'SEAT_NO'] = 0.0

In [570]:
traffic_crashes_people[traffic_crashes_people['SEAT_NO'].isnull()]['PERSON_TYPE'].value_counts()

PEDESTRIAN             21142
BICYCLE                12725
NON-MOTOR VEHICLE       1474
NON-CONTACT VEHICLE      169
Name: PERSON_TYPE, dtype: int64

In [571]:
traffic_crashes_people[traffic_crashes_people['PERSON_TYPE']=='NON-CONTACT VEHICLE'].shape[0]

169

it's right when we find these nan values when the person is PEDESTRIAN or BICYCLE or NON_MOTOR as they are not real vehicles, but what about the NON-CONTACT?, it seems all the values are nan also so not unique value for this type also .
let's fill them all with for example : (-1)

In [572]:
traffic_crashes_people['SEAT_NO'].fillna('-1.0',inplace=True)

CITY and STATE columns are must be related and we may can conclude missing values from each others. but let's see when they being null.

In [573]:
traffic_crashes_people[(traffic_crashes_people['CITY'].isnull()) & (traffic_crashes_people['STATE'].isnull())].shape[0]

453302

They are both nan in most rows , so can't conclude one from each other except some rows .
let's take a deeper look .

In [574]:
traffic_crashes_people['CITY'].value_counts()

CHICAGO          933079
CICERO            12984
SKOKIE             8408
BERWYN             7299
EVANSTON           7228
                  ...  
NAPOLEONVILLE         1
CAMPUS                1
ALLENTON              1
HANCOK                1
BOLINGROOK            1
Name: CITY, Length: 13407, dtype: int64

In [575]:
# df['CITY'] = df['CITY'].str.extract(r'([-+]?\d*\.\d+|\d+)').astype(float)
# df['CITY'].str.extract(r'([-+]?\d*\.\d+|\d+)')

traffic_crashes_people[traffic_crashes_people['CITY'].notnull() & traffic_crashes_people['CITY'].str.contains('\\d',regex=True)]['CITY'].unique()


array(['60655', '60643', '99', 'CHICAG0', '60649', '60615', '60629',
       'GA1', 'OTTAWA, ONTARIO K1S1SQ', 'ELMWO0D', '60614', '60428',
       '60619', '60641', 'LOMBARDY R1 ONTARIO CANADA K0G1L0',
       '9152820310', '606402', '60621', '7430 THUSIS, SWITZERLAND',
       '60076', 'CHICAGO9', '03JUN1991', 'CHICAGO1', 'ALSIP0803', '60645',
       '1B', '60625', 'THAMESFORD RR4, ON', '2', '60628', '1C', '60623',
       'ANNAN RR2', '3CHICAGO', 'CICER0', 'CHI9CAGO', '3', 'CHICAGO0',
       'CHICAG0O', '60636', 'CHICAGO2', '60637', '60632',
       'TORONTO ONTARIO M4N 2C9', 'SUGAR GROVE, IL 60554', '1353',
       'FLOSSMO0R', '1221', 'NORTH YORK ONTARIO CANADA M2J3A5',
       'BRAMPTON, ON L6X0V6', '17915', 'PEKIN61554', 'D624', '60638',
       '75018 PARIS', 'WILLOWBROOK IL 60527', 'WINNIPEG MB R2M 2N6',
       '60647', 'CHICAGO62', 'CHCAG0', '60302', '60529', 'CHI8CAGO',
       'SAUK VILLAGE60411', 'WINNEPEG, MANITOBA,CANADA,ZIP R2J0Z1',
       '60804', 'CHICAGO  60612', 'SEC-4 DWARKA'

??

In [576]:
chicago_mask = traffic_crashes_people['CITY'].str.contains(r'(?i)ch.*i.*c.*a.*g.*o', regex=True)
traffic_crashes_people[traffic_crashes_people['CITY'].notnull() & (chicago_mask)]['CITY'].unique()

array(['CHICAGO', 'CHICAGO RIDGE', 'CHICAGIO', 'EAST CHICAGO',
       'SOUTH CHICAGO HEIGHTS', 'CHICAGO HEIGHTS', 'CHICAGO O',
       'CHICCAGO', 'WEST CHICAGO', 'CHICAGOIL', 'CHICAGO`', 'CHICAGOI',
       'CHICAGO HEIGHT', 'NORTH CHICAGO', 'CHIKCAGO', 'CHICAGFO',
       'CHIICAGO', 'CHCICAGO', 'CHICAGOO', 'CHHICAGO', 'CHICAGOL',
       'CHICASGO', 'CHICGAGO', 'CHICAGO IL', 'CCHICAGO', 'CHICAGGO',
       'CHUICAGO', 'SOUTH CHICAGO', 'CHICAGOQ', 'CHICAGO;', 'CHICAGO N',
       'CHICICAGO', 'CHICAGO-HEIGHT', 'CHICAGO HIEGHTS', 'CHICAQGO',
       'CHICAGHO', 'CHICAGO`I', 'CHIHCAGO', 'CHICAGO AVE', 'CHIOCAGO',
       'CHICAIGO', 'CHICACAGO', 'CHICAGOE', 'SOUTH CHICAGO HEIGHT',
       'SOUTH CHICAGO HEIGHT`', 'CHKICAGO', 'CHICAOGO', 'CHICAAGO',
       'NORHT CHICAGO', 'CHICAFGO', 'SO CHICAGO HEIGHTS', 'EAST  CHICAGO',
       'CHICAGO HIGHTS', 'CHICAGOP', 'CHICVAGO', 'CHICACGO', 'CHICAGO Q',
       'CHIACAGO', 'CHIUCAGO', 'CHICAGO I', 'CHICAHGO', 'CHICAGORIDGE',
       'CHIC AGO', 'CHICAGOQI

In [577]:
traffic_crashes_people.loc[traffic_crashes_people['CITY'].notnull() & (chicago_mask) , 'CITY'] = "CHICAGO"

In [578]:
# chicago_mask = traffic_crashes_people['CITY'].str.contains(r'(?i)\bch(?:i|a|e|y|)[a-z]*g(?:o|a|e|y|)[a-z]*\b', regex=True)
# traffic_crashes_people[traffic_crashes_people['CITY'].notnull() & (chicago_mask)]['CITY'].unique()

In [579]:
chicago_mask = traffic_crashes_people['CITY'].str.contains(r'(?i)\bch(?:i|a|)[a-z]*c(?:a|e|y|)[a-z]*g(?:o|a|e|y)[a-z]*\b', regex=True)
traffic_crashes_people[traffic_crashes_people['CITY'].notnull() & (chicago_mask)]['CITY'].unique()

array(['CHICAGO', 'CHIACGO', 'CHICGO', 'CHCIAGO', 'CHICGAO', 'CHCAGO',
       'CHOCAGO', 'CHCAIGO', 'CHUCAGO', 'CHCIAGO RIDGE', 'CHCIAGO IL',
       'CHICGO HEIGHTS', 'CHICSGO', 'CHICQGO', 'CHJCAGO', 'CHICIGO',
       'CHICCGAO', 'CHCCAGO', 'CHCAAGO', 'CHCIACGO', 'CHCIAGOP',
       'CHICGAO RIDGE', 'CHCAGOQ', 'CHCIAGO HEIGHTS', 'CHICKAMAUGA',
       'CHLCAGO', 'CHICAGA', 'CHKCAGO', 'CHCGAGO', 'CHCIAQGO', 'CHCAHGO',
       'CHIACGO RIDGE', 'NORTH CHCAGO', 'CHCACGO', 'CHICAGE', 'CHHCAGO',
       'CHCAGOI', 'CHACIGO', 'CHCAGO HEIGHTS', 'CHCAOGO', 'CHAICGO',
       'CHACAGO', 'EAST CHCIAGO', 'CHICOGO', 'WEST CHICGO', 'EAST CHCAGO',
       'CHCIAGOL', 'CHCIACOGO', 'CHCVAGO', 'CHICHGO'], dtype=object)

In [580]:
traffic_crashes_people.loc[traffic_crashes_people['CITY'].notnull() & (chicago_mask) , 'CITY'] = "CHICAGO"

In [581]:
# traffic_crashes_people[traffic_crashes_people['CITY'].notnull() & traffic_crashes_people['CITY'].str.startswith('CH')]['CITY'].unique()

In [582]:
import re

In [583]:
chicago_pattern = re.compile(r'C.*H.*I.*C.*A.*G.*O')
def is_chicago(city):
    return bool(chicago_pattern.search(city))
df = traffic_crashes_people[traffic_crashes_people['CITY'].notnull()]
chicago_mask = df['CITY'].str.upper().apply(is_chicago)
df[chicago_mask]['CITY'].unique()

array(['CHICAGO', 'CIHICAGO', 'C HICAGO', 'CVHICAGO', 'CJHICAGO',
       'CDHICAGO', 'CKHICAGO', 'CBHICAGO', 'CG=HICAGO', 'CGHICAGO',
       'CXHICAGO', 'CAHICAGO', 'C;HICAGO', 'CFHICAGO'], dtype=object)

In [584]:
traffic_crashes_people.loc[traffic_crashes_people['CITY'].notnull() & (chicago_mask) , 'CITY'] = "CHICAGO"

In [585]:
chicago_pattern = re.compile(r'[^\w]*C[HICA]*G[^\w]*[O0]', re.IGNORECASE)

def is_chicago(city):
    return bool(chicago_pattern.search(city))
df = traffic_crashes_people[traffic_crashes_people['CITY'].notnull()]
chicago_mask = df['CITY'].str.upper().apply(is_chicago)
df[chicago_mask]['CITY'].unique()

array(['CHICAGO', 'CHGO', 'CICAGO', 'CHIAGO', 'CHIAGO IL', 'CHGO.',
       'CHICAG0', 'CIHCAGO', 'CGICAGO HEIGHTS', 'HICAGO', 'CGICAGO',
       'CJICAGO', 'VHICAGO', 'CHIAGO HEIGHTS', 'CHIAAGO', 'CHGO RIDGE',
       'NORTH CHIAGO', 'HHICAGO', 'SO. CHGO. HEIGHTS', 'CHGO IL',
       'XHICAGO', 'GHICAGO', 'CAGO', 'SO CHGO HEIGHTS', 'WEST CHGO',
       'CHCAG0', 'NORTH CHGO', 'CHGO HTS', 'HCIAGO', 'HCICAGO', 'CHIGO',
       'CHGO1', 'PASCAGOULA', 'CUICAGO', 'SHICAGO', 'CICAGOH', 'CCICAGO',
       'CBICAGO', 'CHCIAG0', 'CKICAGO', 'CHIAGOQ', 'CICAGO.', 'CHAGO',
       'CAHIAGO', 'EAST CHICAG0', 'EAST CGICAGO', 'CJOCAGO', 'CGO'],
      dtype=object)

In [586]:
traffic_crashes_people.loc[traffic_crashes_people['CITY'].notnull() & (chicago_mask) , 'CITY'] = "CHICAGO"

In [587]:
mask = (traffic_crashes_people['CITY'].str.contains('CH',regex=True)
       ) & (traffic_crashes_people['CITY'].str.contains('\\d',regex=True) )
        
traffic_crashes_people[traffic_crashes_people['CITY'].notnull() & (mask) ]['CITY'].unique()


array(['BEECHER /P.O. BOX 29', 'CHICA20', 'CH5CAG6'], dtype=object)

In [588]:
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CH5CAG6' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICA20' , 'CITY'] = "CHICAGO"

In [589]:
traffic_crashes_people[traffic_crashes_people['CITY'].notnull() & traffic_crashes_people['CITY'].str.contains('CHI',regex=True)]['CITY'].unique()


array(['CHICAGO', 'MICHIGAN CITY', 'CHICAO', 'CHICAHO', 'CHICAGP',
       'SCHILLER PARK', 'CHICAGI', 'CHICAG', 'CHICADO', 'CHI', 'CHICAFO',
       'CHI. RIDGE', 'CHIGACO', 'SCHILLER PK', 'WICHITA', 'MICHIGAN',
       'CHICAOG', 'CHIVAGO', 'CHICAGL', 'CHICACO', 'WICHITA FALLS',
       'CHICAGIL', 'SCHILLIER PARK', 'CHIC', 'NORTH CHICAGI', 'CHICAGG',
       'NORTH CHICAO', 'CHIACHO', 'SCHILLER TER', 'CHICA', 'CHICAOO',
       'CHIGAO HEIGHTS', 'CHIXAGO', 'CHILLICOTHE', 'HUTCHINSON',
       'WAXAHACHIE', 'CHICAHO HEIGHTS', 'SCHILLERPARK', 'CHILLER PARK',
       'CHICO', 'SCHILLER', 'CHICAG HEIGHTS', 'SCHILLER APRK',
       'AICHI-JAPAN', 'WEST CHICAGP', 'CHIGAGO', 'CHICAQO', 'PELAHATCHIE',
       'PRAIRIE DU CHIEN', 'SCHILER PARK', 'CHIPPEWA FALLS', 'CHICAHP',
       'SCHILLEER PARK', 'CHICKEN', 'MICHIAGN CITY', 'SCHILLAR PARK',
       'EAST CHICAO', 'KEATCHIE', 'ATCHISON', 'MICHIANA SHORE', 'CHICOAG',
       'MICHIGAN SHORES', 'CHIC GAO', 'WITCHITA', 'SCHILLER PRK',
       'MICHIANA SHO

In [590]:
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICAGI' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICAGP' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICAHO' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICAO' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHIDAGO' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHIACO' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICG' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICCAG' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICAOG' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHIAVGO' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHIACHO' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICOAG' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'NORTH CHICAGI' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICICAG' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'EAST CHIGAGO' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHIGAGO' , 'CITY'] = "CHICAGO"
traffic_crashes_people.loc[traffic_crashes_people['CITY'] ==  'CHICAQO' , 'CITY'] = "CHICAGO"

after all the regular expressions i tried to type, still have alot of misstyped CHICAGO in CITY column.
so i had to replace some of them manually as above, but still i time consuming and bad treatment, so let's see what we will do with this column at all.

In [591]:
traffic_crashes_people['CITY'].value_counts()


CHICAGO        952028
CICERO          12984
SKOKIE           8408
BERWYN           7299
EVANSTON         7228
                ...  
CAMPUS              1
HANCOK              1
WINFIELD DR         1
PHILADLPHIA         1
OTHELLO             1
Name: CITY, Length: 13073, dtype: int64

In [592]:
traffic_crashes_people.groupby('CITY').agg({'PERSON_ID' : 'count'})

PERSON_ID
CITY                       
,HARWOOD HEIGHTS          1
.BRIDGEVIEW               1
03JUN1991                 1
0AK PARK                  1
1                         2
...                     ...
\MOUNT PROSPECT           1
\WESTCHESTER              1
`EVERGREEN PARK           1
`FLOSSMOOR                1
`RIVERDALE                1

[13073 rows x 1 columns]

In [593]:
traffic_crashes_people[traffic_crashes_people['CITY'] == '.BRIDGEVIEW']

PERSON_ID PERSON_TYPE  \
708551    O29889      DRIVER   

                                          CRASH_RECORD_ID  VEHICLE_ID  \
708551  8b80e7852289b7520ef5239e0e74158a5809b304629688...     28861.0   

                    CRASH_DATE SEAT_NO         CITY STATE ZIPCODE SEX   AGE  \
708551  02/20/2016 02:00:00 PM     0.0  .BRIDGEVIEW    IL   60455   M  47.0   

       DRIVERS_LICENSE_STATE DRIVERS_LICENSE_CLASS SAFETY_EQUIPMENT  \
708551                    IL                     A    USAGE UNKNOWN   

           AIRBAG_DEPLOYED EJECTION    INJURY_CLASSIFICATION HOSPITAL  \
708551  DEPLOYMENT UNKNOWN     NONE  NO INDICATION OF INJURY      NaN   

       EMS_AGENCY EMS_RUN_NO DRIVER_ACTION DRIVER_VISION PHYSICAL_CONDITION  \
708551        NaN        NaN       UNKNOWN  NOT OBSCURED             NORMAL   

       PEDPEDAL_ACTION PEDPEDAL_VISIBILITY PEDPEDAL_LOCATION  \
708551             NaN                 NaN               NaN   

              BAC_RESULT  BAC_RESULT VALUE CELL_PHONE_USE  
708551  TEST NOT OFFERED               NaN            NaN

let's expect the real format of the .BRIDGEVIEW : 

In [594]:
traffic_crashes_people[traffic_crashes_people['CITY'] == 'BRIDGEVIEW'].shape[0]

1807

Exactlyy.. i now have alot of noise and for each city i have in my CITY column maybe , so i have to try to find a global regualr expression to clean the names of cities. 

In [595]:
def clean_city_name(city):
       # Remove leading and trailing special characters, spaces, and digits
       city = re.sub(r'^[^a-zA-Z]+|[^a-zA-Z]+$', '', city)
       
       # Convert to lowercase and remove extra spaces
       city = ' '.join(city.split())
       
       return city


    
# df = traffic_crashes_people[traffic_crashes_people['CITY'].notnull()][['CITY','PERSON_ID']].copy()   
    
traffic_crashes_people['CITY'] = traffic_crashes_people[traffic_crashes_people['CITY'].notnull()]['CITY'].apply(clean_city_name)

In [596]:
traffic_crashes_people[traffic_crashes_people['CITY'] == ".BRIDGEVIEW"]

Empty DataFrame
Columns: [PERSON_ID, PERSON_TYPE, CRASH_RECORD_ID, VEHICLE_ID, CRASH_DATE, SEAT_NO, CITY, STATE, ZIPCODE, SEX, AGE, DRIVERS_LICENSE_STATE, DRIVERS_LICENSE_CLASS, SAFETY_EQUIPMENT, AIRBAG_DEPLOYED, EJECTION, INJURY_CLASSIFICATION, HOSPITAL, EMS_AGENCY, EMS_RUN_NO, DRIVER_ACTION, DRIVER_VISION, PHYSICAL_CONDITION, PEDPEDAL_ACTION, PEDPEDAL_VISIBILITY, PEDPEDAL_LOCATION, BAC_RESULT, BAC_RESULT VALUE, CELL_PHONE_USE]
Index: []

In [597]:
traffic_crashes_people[traffic_crashes_people['CITY'] == "BRIDGEVIEW"].shape[0]

1809

We see that we have more two values of BRIDGEVIEW city in the column instead of noisy ones. so not bad.

In [598]:
traffic_crashes_people.groupby('CITY').agg({'PERSON_ID' : 'count'})

PERSON_ID
CITY                 
                  451
A                   2
A GRANGE            1
A;BIA               1
AACWORTH            1
...               ...
ZIONZVILLE          1
ZITACUARO           1
ZUBBACK             1
ZURICH              2
ZURORA              1

[12822 rows x 1 columns]

In [599]:
traffic_crashes_people['STATE'].unique()

array(['IL', nan, 'LA', 'IN', 'XX', 'WI', 'KY', 'CT', 'NY', 'WA', 'CA',
       'MI', 'AL', 'GA', 'NV', 'PA', 'AR', 'SC', 'IA', 'OH', 'TX', 'MO',
       'NC', 'FL', 'CO', 'NJ', 'MN', 'NB', 'OK', 'TN', 'AZ', 'VA', 'MT',
       'NM', 'DC', 'DE', 'WV', 'MD', 'SD', 'KS', 'ID', 'MA', 'UT', 'MS',
       'ME', 'WY', 'HI', 'NH', 'OR', 'ND', 'RI', 'VT', 'AK'], dtype=object)

In [600]:
traffic_crashes_people[traffic_crashes_people['CITY'] == "CHICAGO"]['STATE'].value_counts()

IL    940184
IN      1322
ID       424
LA        22
HI        11
XX         6
IA         6
OH         4
WI         4
KS         4
CA         3
FL         3
MS         2
MI         2
GA         2
AL         2
ME         2
PA         1
AK         1
TX         1
MO         1
NB         1
MN         1
Name: STATE, dtype: int64

In [601]:
traffic_crashes_people['CITY'].value_counts()

CHICAGO          952028
CICERO            12986
SKOKIE             8410
BERWYN             7301
EVANSTON           7228
                  ...  
WINNEKA               1
WESTERFIELD           1
GAITHERSBURGH         1
OAKAWN                1
BLUE ISAND            1
Name: CITY, Length: 12822, dtype: int64

After all tries and attempts , still alot and alot of noise in city names and chaosity, still the same CITY belongs to many different states and the vice-versa. 
Let's assign the most frequent value in CITY ('CHICAGO') and STATE('IL') to the nan and missed values in these columns.

In [602]:
traffic_crashes_people.loc[traffic_crashes_people['CITY'] == '', 'CITY'] = 'CHICAGO'

In [603]:
traffic_crashes_people.loc[traffic_crashes_people['CITY'].isnull() , 'CITY'] = 'CHICAGO'

In [604]:
traffic_crashes_people.loc[traffic_crashes_people['STATE'].isnull() , 'STATE'] = 'IL'

#### what about ZIPCODE :

In [605]:
traffic_crashes_people['ZIPCODE'].nunique()

14401

In [606]:
traffic_crashes_people[traffic_crashes_people['CITY'] == 'CHICAGO']['ZIPCODE'].unique()

array(['60651', '60620', nan, ..., '66289', '53218', '60336'],
      dtype=object)

In [607]:
traffic_crashes_people[traffic_crashes_people['CITY'] == 'CICERO']['ZIPCODE'].unique()

array(['60804', nan, '60604', '60644', '60650', '60864', '90804', '60622',
       '`60804', '60806', '60004', '60608', '60904', '60684', '60648',
       '60402', '60504', '60808', '60408', '60184', '60801', '60084',
       '60302', '60639', '60809', '60304', '06084', '60802', '60641',
       '60454', '60623', '60814', '60834', '60647', '80804', '60884',
       '60618', '60613', '60480', '60204', '60612', '60064', '60805',
       '60804-1144', '60632', '60707', '60803', '62804', '60824', '13039',
       '60415', '60955', '60844', '60680', '60506', '6004', '60609',
       '60638', '600804', '60104', '60651', '69804', '60850', '60840',
       '50804', '60130', '60201', 'UNKNWON', 'UNKNOWN', '66804', '60646'],
      dtype=object)

In [608]:
traffic_crashes_people[traffic_crashes_people['STATE'] == 'IL']['ZIPCODE'].unique()

array(['60651', '60620', nan, ..., '66289', '55106', '60336'],
      dtype=object)

We see there is no realtion between the ZIPCODE and the CITY or the STATE , there is even a clear range. 
simply let's drop the whole column .

In [609]:
traffic_crashes_people = traffic_crashes_people.drop('ZIPCODE', axis=1)

#### again a large miss in SEX column .

In [610]:
traffic_crashes_people['SEX'].isnull().sum()

29939

In [611]:
traffic_crashes_people['SEX'].value_counts()

M    934343
F    679200
X    164287
Name: SEX, dtype: int64

may the X represent that is unknown?! but how ? let's replace them also as we fill nan values . 

In [612]:
male_count = traffic_crashes_people['SEX'].eq('M').sum()
female_count = traffic_crashes_people['SEX'].eq('F').sum()
total_count = male_count + female_count
male_ratio = male_count / total_count
female_ratio = female_count / total_count
print(male_ratio)
print(female_ratio)

0.579062968882763
0.420937031117237


In [613]:
missing_indices = traffic_crashes_people[traffic_crashes_people['SEX'] == 'X'].index
fill_values = np.random.choice(['M', 'F'], size=len(missing_indices), p=[male_ratio, female_ratio])
traffic_crashes_people.loc[missing_indices, 'SEX'] = fill_values

In [614]:
traffic_crashes_people['SEX'].value_counts()

M    1029193
F     748637
Name: SEX, dtype: int64

We replaced the X rows with M or F with respect to its proportion existing through the dataset. 
and we will do the same to fill NaNs.

In [615]:
missing_indices = traffic_crashes_people[traffic_crashes_people['SEX'].isnull()].index
fill_values = np.random.choice(['M', 'F'], size=len(missing_indices), p=[male_ratio, female_ratio])
traffic_crashes_people.loc[missing_indices, 'SEX'] = fill_values

In [616]:
traffic_crashes_people['SEX'].isnull().sum()

0

In [617]:
traffic_crashes_people['AGE'].unique()

array([ 25.,  37.,  nan,  34.,  53.,  65.,  26.,  64.,  33.,  22.,  70.,
        51.,  61.,  28.,  27.,  29.,  58.,  32.,  18.,  48.,  30.,  43.,
        24.,  57.,  23.,  52.,  36.,  47.,   0.,  31.,  42.,  39.,  60.,
        35.,  44.,  54.,  63.,  49.,  40.,  45.,  62.,  50.,  71.,  21.,
        46.,  59.,  67.,  68.,  74.,  56.,  38.,  83.,  19.,  16.,  75.,
        55.,  13.,  80.,  69.,  86.,  78.,  20.,  41.,  72.,  66.,  17.,
        73.,  81.,  15.,  82.,  14.,  92.,  76.,   3.,  79.,   5.,   1.,
        10.,  77.,   6.,   2.,   9.,   7.,  11.,   8.,  12.,  97.,   4.,
        84.,  90.,  89.,  85.,  87.,  88.,  91.,  99., 104.,  98.,  94.,
        93., 100.,  -1., -47.,  95., 102., 109., 107., 103.,  96., 101.,
       110., 108., -49., 105., -40., 106., -59.])

##### wait we have negative ages ?! 
and what about the underage drivers ?!

In [618]:
traffic_crashes_people[traffic_crashes_people['AGE'] < 0].shape

(10, 28)

In [619]:
traffic_crashes_people = traffic_crashes_people[(traffic_crashes_people['AGE'] >= 0) | (traffic_crashes_people['AGE'].isnull())]


In [620]:
traffic_crashes_people[traffic_crashes_people['AGE'] < 0].shape

(0, 28)

In [621]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == "DRIVER" ) & (traffic_crashes_people['AGE'] < 10)].shape[0]

9771

it's not allowed for the person under 10 to drive yes, but he/she able to. 
people in age 10 and more may able to drive so everything is possible, so we put 10 yrs old our threshold and all drivers with age underthat we replace them with the whole mean of ages.

In [622]:
traffic_crashes_people['AGE'].mean()

37.90324803095788

In [623]:
traffic_crashes_people.loc[(traffic_crashes_people['PERSON_TYPE'] == "DRIVER" ) & (traffic_crashes_people['AGE'] < 10),'AGE'] = traffic_crashes_people['AGE'].mean()


In [624]:
traffic_crashes_people['AGE'].isnull().sum()

527568

In [625]:
stat = traffic_crashes_people[['SEX','AGE']].groupby('SEX').agg({'AGE':'mean'})
stat

AGE
SEX           
F    37.047884
M    39.039594

that the mean of age of each gender , we see they are so close to each other.
let's see the mean of age of people involved in accidents between 12am and 6am.

In [626]:
traffic_crashes_people[(pd.to_datetime(traffic_crashes_people['CRASH_DATE']).dt.hour >= 0 ) &
          ( pd.to_datetime(traffic_crashes_people['CRASH_DATE']).dt.hour < 6 ) 
         ]['AGE'].mean()

34.39443984380709

In [627]:
stat['AGE'][0] ##the females mean age 

37.047883808524276

again still a close value to others means. 
so let's fill the nan ages with mean value for each gender. 

In [628]:
traffic_crashes_people.loc[
    (traffic_crashes_people['SEX'] == 'M') & traffic_crashes_people['AGE'].isnull(),
    'AGE'] = stat['AGE'][1]
traffic_crashes_people.loc[
    (traffic_crashes_people['SEX'] == 'F') & traffic_crashes_people['AGE'].isnull(),
    'AGE'] = stat['AGE'][0]

In [629]:
traffic_crashes_people['AGE'].isnull().sum()

0

### DRIVER_LISENCE_ class and state 

In [630]:
traffic_crashes_people['DRIVERS_LICENSE_STATE'].value_counts()

IL    970216
XX     23704
IN     17990
WI      5903
MI      4521
       ...  
DH         1
CI         1
GM         1
GP         1
LU         1
Name: DRIVERS_LICENSE_STATE, Length: 210, dtype: int64

In [631]:
traffic_crashes_people.loc[traffic_crashes_people['DRIVERS_LICENSE_STATE'].isnull() , 'DRIVERS_LICENSE_STATE'] = 'IL'

In [632]:
traffic_crashes_people['DRIVERS_LICENSE_CLASS'].value_counts()

D     770951
A      31578
C      24195
B      24101
DM     13268
       ...  
GV         1
TA         1
F1         1
BV         1
LT         1
Name: DRIVERS_LICENSE_CLASS, Length: 281, dtype: int64

In [633]:
traffic_crashes_people.loc[traffic_crashes_people['DRIVERS_LICENSE_CLASS'].isnull() , 'DRIVERS_LICENSE_CLASS'] = 'D'

### Another new column

In [634]:
traffic_crashes_people['SAFETY_EQUIPMENT'].value_counts()

SAFETY BELT USED                              851445
USAGE UNKNOWN                                 847531
NONE PRESENT                                   60064
SAFETY BELT NOT USED                            9903
HELMET NOT USED                                 8411
CHILD RESTRAINT USED                            7822
CHILD RESTRAINT - FORWARD FACING                4425
BICYCLE HELMET (PEDACYCLIST INVOLVED ONLY)      3256
CHILD RESTRAINT - TYPE UNKNOWN                  2329
CHILD RESTRAINT - REAR FACING                   2045
HELMET USED                                     1358
DOT COMPLIANT MOTORCYCLE HELMET                 1295
BOOSTER SEAT                                    1166
CHILD RESTRAINT NOT USED                         808
SHOULD/LAP BELT USED IMPROPERLY                  239
NOT DOT COMPLIANT MOTORCYCLE HELMET              236
WHEELCHAIR                                       213
CHILD RESTRAINT USED IMPROPERLY                  146
STRETCHER                                     

In [635]:
traffic_crashes_people['SAFETY_EQUIPMENT'].isnull().sum()

5027

In [636]:
traffic_crashes_people['EJECTION'].isnull().sum()

22419

In [637]:
def get_mode(series):
    return series.mode().iloc[0] if not series.empty else None

result = traffic_crashes_people.groupby('EJECTION')['SAFETY_EQUIPMENT'].agg(get_mode).reset_index()
result

EJECTION  SAFETY_EQUIPMENT
0                NONE  SAFETY BELT USED
1   PARTIALLY EJECTED  SAFETY BELT USED
2     TOTALLY EJECTED   HELMET NOT USED
3  TRAPPED/EXTRICATED     USAGE UNKNOWN
4             UNKNOWN     USAGE UNKNOWN

In [638]:
result = traffic_crashes_people.groupby('SAFETY_EQUIPMENT')['EJECTION'].agg(get_mode).reset_index()
result

SAFETY_EQUIPMENT EJECTION
0   BICYCLE HELMET (PEDACYCLIST INVOLVED ONLY)     NONE
1                                 BOOSTER SEAT     NONE
2             CHILD RESTRAINT - FORWARD FACING     NONE
3                CHILD RESTRAINT - REAR FACING     NONE
4               CHILD RESTRAINT - TYPE UNKNOWN     NONE
5                     CHILD RESTRAINT NOT USED     NONE
6                         CHILD RESTRAINT USED     NONE
7              CHILD RESTRAINT USED IMPROPERLY     NONE
8              DOT COMPLIANT MOTORCYCLE HELMET     NONE
9                              HELMET NOT USED     NONE
10                                 HELMET USED     NONE
11                                NONE PRESENT     NONE
12         NOT DOT COMPLIANT MOTORCYCLE HELMET     NONE
13                        SAFETY BELT NOT USED     NONE
14                            SAFETY BELT USED     NONE
15             SHOULD/LAP BELT USED IMPROPERLY     NONE
16                                   STRETCHER     NONE
17                               USAGE UNKNOWN     NONE
18                                  WHEELCHAIR     NONE

In [639]:
traffic_crashes_people[traffic_crashes_people['SAFETY_EQUIPMENT'] == 'SAFETY BELT USED']['EJECTION'].value_counts()

NONE                  844826
UNKNOWN                 4533
TOTALLY EJECTED         1204
PARTIALLY EJECTED        374
TRAPPED/EXTRICATED       313
Name: EJECTION, dtype: int64

In [640]:
result = traffic_crashes_people.groupby('INJURY_CLASSIFICATION')['SAFETY_EQUIPMENT'].agg(get_mode).reset_index()
result

INJURY_CLASSIFICATION  SAFETY_EQUIPMENT
0                     FATAL     USAGE UNKNOWN
1     INCAPACITATING INJURY     USAGE UNKNOWN
2   NO INDICATION OF INJURY     USAGE UNKNOWN
3  NONINCAPACITATING INJURY  SAFETY BELT USED
4     REPORTED, NOT EVIDENT  SAFETY BELT USED

In [641]:
result = traffic_crashes_people.groupby('EJECTION')['INJURY_CLASSIFICATION'].agg(get_mode).reset_index()
result

EJECTION    INJURY_CLASSIFICATION
0                NONE  NO INDICATION OF INJURY
1   PARTIALLY EJECTED  NO INDICATION OF INJURY
2     TOTALLY EJECTED  NO INDICATION OF INJURY
3  TRAPPED/EXTRICATED  NO INDICATION OF INJURY
4             UNKNOWN  NO INDICATION OF INJURY

###### We can use each column of these three columns (EJECTION , INJURY_CLASSIFICATION , SAFETY_EQUIPMENT) to guess the most-frequent value in the other columns for each value.  
- the most-frequent for 'NONE' in EJECTION is SAFETY BELT USED and the vice-versa.

- but we see that for every value in SAFETY_EQUIPMENT the most frequent in corresponding value in EJECTION column is NONE.

- when 'SAFETY BELT USED' it has a 845135 values with NONE in 'EJECTION' column and it is very far value from the others.   but that's not happen when we talk about other values in 'SAFETY_EQUIPMENT' it is varries closely between 'EJECTION' values. so it is logical to assign nan values in both columns with SAFETY BELT USED and NONE when they met so many times , but not logical to give the value NONE for the EJECTION through all different values in 'SAFETY_EQUIPMENT' when the values of appearing is nearlly equal. 

In [642]:
traffic_crashes_people.loc[
    (traffic_crashes_people['EJECTION'] == 'NONE') & (traffic_crashes_people['SAFETY_EQUIPMENT'].isnull()),
    'SAFETY_EQUIPMENT'] = 'SAFETY BELT USED'

traffic_crashes_people.loc[
    (traffic_crashes_people['SAFETY_EQUIPMENT'] == 'SAFETY BELT USED') & (traffic_crashes_people['EJECTION'].isnull()),
    'EJECTION'] = 'NONE'

In [643]:
traffic_crashes_people['SAFETY_EQUIPMENT'].isnull().sum()

4837

In [644]:
traffic_crashes_people['EJECTION'].isnull().sum()

22224

In [645]:
traffic_crashes_people[traffic_crashes_people['PERSON_TYPE'] == 'BICYCLE']['EJECTION'].value_counts()

NONE                  8178
TOTALLY EJECTED       2073
UNKNOWN               1541
PARTIALLY EJECTED      488
TRAPPED/EXTRICATED      21
Name: EJECTION, dtype: int64

In [646]:
traffic_crashes_people[traffic_crashes_people['INJURY_CLASSIFICATION'] == 'FATAL']['EJECTION'].value_counts()

NONE                  447
TOTALLY EJECTED       114
UNKNOWN                89
TRAPPED/EXTRICATED     63
PARTIALLY EJECTED      10
Name: EJECTION, dtype: int64

In [647]:
traffic_crashes_people[traffic_crashes_people['INJURY_CLASSIFICATION'] == 'INCAPACITATING INJURY']['EJECTION'].value_counts()

NONE                  10775
UNKNOWN                1020
TOTALLY EJECTED         819
TRAPPED/EXTRICATED      188
PARTIALLY EJECTED        87
Name: EJECTION, dtype: int64

the EJECTION couldn't be filled with more special values , as we see it is not a condition to be the person ejected when the INJURY_CLASSIFICATION is FATAL or INCAPACITATING INJURY , not even for when PERSON_TYPE is BICYCLE or not , most of bicyclists are not ejected , so let's fill the rest nan values with UNKOWN , which it already a categgory in this column .

In [648]:
traffic_crashes_people = traffic_crashes_people.fillna({'EJECTION' : 'UNKNOWN'})

In [649]:
traffic_crashes_people['INJURY_CLASSIFICATION'].value_counts()

NO INDICATION OF INJURY     1649971
NONINCAPACITATING INJURY      88532
REPORTED, NOT EVIDENT         51117
INCAPACITATING INJURY         16456
FATAL                           994
Name: INJURY_CLASSIFICATION, dtype: int64

In [650]:
traffic_crashes_people['INJURY_CLASSIFICATION'].isnull().sum()

689

In [651]:
traffic_crashes_people[traffic_crashes_people['INJURY_CLASSIFICATION'].isnull()]['EJECTION'].unique()

array(['UNKNOWN', 'NONE'], dtype=object)

when i have a nan value in INJURY_CLASSIFICATION all EJECTION values are in NONE or UNKNOWN , don't have real guide of the person is ejected or not . 

In [652]:
result = traffic_crashes_people.groupby('SAFETY_EQUIPMENT')['INJURY_CLASSIFICATION'].agg(get_mode).reset_index()
result

SAFETY_EQUIPMENT     INJURY_CLASSIFICATION
0   BICYCLE HELMET (PEDACYCLIST INVOLVED ONLY)  NONINCAPACITATING INJURY
1                                 BOOSTER SEAT   NO INDICATION OF INJURY
2             CHILD RESTRAINT - FORWARD FACING   NO INDICATION OF INJURY
3                CHILD RESTRAINT - REAR FACING   NO INDICATION OF INJURY
4               CHILD RESTRAINT - TYPE UNKNOWN   NO INDICATION OF INJURY
5                     CHILD RESTRAINT NOT USED   NO INDICATION OF INJURY
6                         CHILD RESTRAINT USED   NO INDICATION OF INJURY
7              CHILD RESTRAINT USED IMPROPERLY   NO INDICATION OF INJURY
8              DOT COMPLIANT MOTORCYCLE HELMET   NO INDICATION OF INJURY
9                              HELMET NOT USED  NONINCAPACITATING INJURY
10                                 HELMET USED  NONINCAPACITATING INJURY
11                                NONE PRESENT   NO INDICATION OF INJURY
12         NOT DOT COMPLIANT MOTORCYCLE HELMET   NO INDICATION OF INJURY
13                        SAFETY BELT NOT USED   NO INDICATION OF INJURY
14                            SAFETY BELT USED   NO INDICATION OF INJURY
15             SHOULD/LAP BELT USED IMPROPERLY   NO INDICATION OF INJURY
16                                   STRETCHER   NO INDICATION OF INJURY
17                               USAGE UNKNOWN   NO INDICATION OF INJURY
18                                  WHEELCHAIR   NO INDICATION OF INJURY

In [653]:
traffic_crashes_people[traffic_crashes_people['INJURY_CLASSIFICATION'].isnull()]['SAFETY_EQUIPMENT'].unique()

array([nan, 'USAGE UNKNOWN', 'SAFETY BELT USED', 'NONE PRESENT',
       'NOT DOT COMPLIANT MOTORCYCLE HELMET'], dtype=object)

In [654]:
traffic_crashes_people[traffic_crashes_people['INJURY_CLASSIFICATION'] == 'FATAL']['SAFETY_EQUIPMENT'].value_counts()

USAGE UNKNOWN                                 469
NONE PRESENT                                  198
SAFETY BELT USED                               86
SAFETY BELT NOT USED                           67
HELMET NOT USED                                60
DOT COMPLIANT MOTORCYCLE HELMET                22
HELMET USED                                     9
BICYCLE HELMET (PEDACYCLIST INVOLVED ONLY)      4
CHILD RESTRAINT NOT USED                        3
NOT DOT COMPLIANT MOTORCYCLE HELMET             3
CHILD RESTRAINT USED                            1
CHILD RESTRAINT - TYPE UNKNOWN                  1
WHEELCHAIR                                      1
Name: SAFETY_EQUIPMENT, dtype: int64

the most-frequent of different SAFETY_EQUIPMENT values is always NO INDICATION OF INJURY , and as we see the values always distribute randomly. so we can't conclude any specific values for INJURY. we fill them with UKNOWN.

In [655]:
traffic_crashes_people['INJURY_CLASSIFICATION'].fillna('UNKNOWN',inplace=True)

In [656]:
result = traffic_crashes_people.groupby('EJECTION')['SAFETY_EQUIPMENT'].agg(get_mode).reset_index()
result

EJECTION  SAFETY_EQUIPMENT
0                NONE  SAFETY BELT USED
1   PARTIALLY EJECTED  SAFETY BELT USED
2     TOTALLY EJECTED   HELMET NOT USED
3  TRAPPED/EXTRICATED     USAGE UNKNOWN
4             UNKNOWN     USAGE UNKNOWN

In [657]:
result = traffic_crashes_people.groupby('INJURY_CLASSIFICATION')['SAFETY_EQUIPMENT'].agg(get_mode).reset_index()
result

INJURY_CLASSIFICATION  SAFETY_EQUIPMENT
0                     FATAL     USAGE UNKNOWN
1     INCAPACITATING INJURY     USAGE UNKNOWN
2   NO INDICATION OF INJURY     USAGE UNKNOWN
3  NONINCAPACITATING INJURY  SAFETY BELT USED
4     REPORTED, NOT EVIDENT  SAFETY BELT USED
5                   UNKNOWN     USAGE UNKNOWN

In [658]:
traffic_crashes_people.loc[
    (traffic_crashes_people['INJURY_CLASSIFICATION'] == 'NONINCAPACITATING INJURY') & (traffic_crashes_people['SAFETY_EQUIPMENT'].isnull()),
    'SAFETY_EQUIPMENT'] = 'SAFETY BELT USED'

In [659]:
traffic_crashes_people.loc[
    (traffic_crashes_people['INJURY_CLASSIFICATION'] == 'REPORTED, NOT EVIDENT') & (traffic_crashes_people['SAFETY_EQUIPMENT'].isnull()),
    'SAFETY_EQUIPMENT'] = 'SAFETY BELT USED'

In [660]:
traffic_crashes_people['SAFETY_EQUIPMENT'].isnull().sum()

2131

In [661]:
traffic_crashes_people['SAFETY_EQUIPMENT'].fillna('USAGE UNKNOWN',inplace=True)

## Let's see where we reach in data cleaning

In [662]:
info_df = pd.DataFrame()
info_df['missing_val'] = traffic_crashes_peoplerashes_people.isnull().sum()
info_df['missing_val_ratio'] = (info_df['missing_val'] / traffic_crashes_people.shape[0] * 100).round().astype(int)
info_df

missing_val  missing_val_ratio
PERSON_ID                        0                  0
PERSON_TYPE                      0                  0
CRASH_RECORD_ID                  0                  0
VEHICLE_ID                       0                  0
CRASH_DATE                       0                  0
SEAT_NO                          0                  0
CITY                             0                  0
STATE                            0                  0
SEX                              0                  0
AGE                              0                  0
DRIVERS_LICENSE_STATE            0                  0
DRIVERS_LICENSE_CLASS            0                  0
SAFETY_EQUIPMENT                 0                  0
AIRBAG_DEPLOYED              34761                  2
EJECTION                         0                  0
INJURY_CLASSIFICATION            0                  0
HOSPITAL                   1508729                 83
EMS_AGENCY                 1622752                 90
EMS_RUN_NO                 1777039                 98
DRIVER_ACTION               368780                 20
DRIVER_VISION               369324                 20
PHYSICAL_CONDITION          367795                 20
PEDPEDAL_ACTION            1773341                 98
PEDPEDAL_VISIBILITY        1773404                 98
PEDPEDAL_LOCATION          1773347                 98
BAC_RESULT                  367809                 20
BAC_RESULT VALUE           1805739                100
CELL_PHONE_USE             1806600                100

In [663]:
result = traffic_crashes_people.groupby('INJURY_CLASSIFICATION')['AIRBAG_DEPLOYED'].agg(get_mode).reset_index()
result

INJURY_CLASSIFICATION        AIRBAG_DEPLOYED
0                     FATAL  DEPLOYED, COMBINATION
1     INCAPACITATING INJURY         DID NOT DEPLOY
2   NO INDICATION OF INJURY         DID NOT DEPLOY
3  NONINCAPACITATING INJURY         DID NOT DEPLOY
4     REPORTED, NOT EVIDENT         DID NOT DEPLOY
5                   UNKNOWN     DEPLOYMENT UNKNOWN

In [664]:
traffic_crashes_people[traffic_crashes_people['INJURY_CLASSIFICATION'] == 'FATAL']['AIRBAG_DEPLOYED'].value_counts()

DEPLOYED, COMBINATION                     207
DEPLOYED, FRONT                           138
NOT APPLICABLE                            123
DID NOT DEPLOY                            109
DEPLOYMENT UNKNOWN                         95
DEPLOYED, SIDE                             18
DEPLOYED OTHER (KNEE, AIR, BELT, ETC.)      2
Name: AIRBAG_DEPLOYED, dtype: int64

In [665]:
result = traffic_crashes_people.groupby('EJECTION')['AIRBAG_DEPLOYED'].agg(get_mode).reset_index()
result

EJECTION     AIRBAG_DEPLOYED
0                NONE      DID NOT DEPLOY
1   PARTIALLY EJECTED      DID NOT DEPLOY
2     TOTALLY EJECTED      DID NOT DEPLOY
3  TRAPPED/EXTRICATED  DEPLOYMENT UNKNOWN
4             UNKNOWN  DEPLOYMENT UNKNOWN

it's randomly distribution between calsses in three columns , no real strong meaningful realtion.

In [666]:
# traffic_crashes_people = traffic_crashes_people.fillna({'AIRBAG_DEPLOYED' : 'DEPLOYMENT UNKNOWN'})
traffic_crashes_people['AIRBAG_DEPLOYED'].fillna('DEPLOYMENT UNKNOWN',inplace=True)

drop the not necessary columns with very loss in the data. 

In [667]:
traffic_crashes_people = traffic_crashes_people.drop('HOSPITAL', axis=1)
traffic_crashes_people = traffic_crashes_peoplerashes_people.drop('EMS_AGENCY', axis=1)
traffic_crashes_people = traffic_crashes_people.drop('EMS_RUN_NO', axis=1)

In [668]:
traffic_crashes_peoplerashes_people['DRIVER_ACTION'].value_counts()

NONE                                 514326
UNKNOWN                              362648
FAILED TO YIELD                      131120
OTHER                                128768
FOLLOWED TOO CLOSELY                  85256
IMPROPER BACKING                      42737
IMPROPER TURN                         37738
IMPROPER LANE CHANGE                  36940
IMPROPER PASSING                      32137
DISREGARDED CONTROL DEVICES           25392
TOO FAST FOR CONDITIONS               21632
WRONG WAY/SIDE                         5764
IMPROPER PARKING                       5307
OVERCORRECTED                          2682
EVADING POLICE VEHICLE                 2296
CELL PHONE USE OTHER THAN TEXTING      2118
EMERGENCY VEHICLE ON CALL              1319
TEXTING                                 567
STOPPED SCHOOL BUS                      172
LICENSE RESTRICTIONS                     60
Name: DRIVER_ACTION, dtype: int64

In [669]:
traffic_crashes_people['DRIVER_ACTION'].isnull().sum()

368780

As we serach in an atrribute that belongs only to drivers. so let's add a new value for DRIVER_ACTION as NOT RELEVANT when the person is pedetrian or passenger or bicyclist.

In [670]:
traffic_crashes_people[(traffic_crashes_peoplerashes_people['PERSON_TYPE'].isin(['PASSENGER','PEDESTRIAN','BICYCLE'])) & (traffic_crashes_people['DRIVER_ACTION'].isnull())].shape[0]

368206

In [671]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'] == 'NON_MOTOR VEHICLE') & (traffic_crashes_people['DRIVER_ACTION'].isnull())].shape[0]

0

In [672]:
traffic_crashes_people.loc[
    (traffic_crashes_people['PERSON_TYPE'].isin(['PASSENGER','PEDESTRIAN','BICYCLE'])) & (traffic_crashes_people['DRIVER_ACTION'].isnull()),
    'DRIVER_ACTION'] = 'NOT RELEVANT'

In [673]:
traffic_crashes_people['DRIVER_ACTION'].isnull().sum()

574

In [674]:
traffic_crashes_people['DRIVER_ACTION'].unique()

array(['UNKNOWN', 'NONE', 'IMPROPER BACKING', 'IMPROPER TURN',
       'DISREGARDED CONTROL DEVICES', 'TOO FAST FOR CONDITIONS', 'OTHER',
       'FOLLOWED TOO CLOSELY', 'IMPROPER LANE CHANGE', 'IMPROPER PASSING',
       'FAILED TO YIELD', 'NOT RELEVANT', 'EVADING POLICE VEHICLE',
       'OVERCORRECTED', 'WRONG WAY/SIDE', 'TEXTING', 'IMPROPER PARKING',
       'EMERGENCY VEHICLE ON CALL', 'CELL PHONE USE OTHER THAN TEXTING',
       nan, 'STOPPED SCHOOL BUS', 'LICENSE RESTRICTIONS'], dtype=object)

In [675]:
traffic_crashes_people[(traffic_crashes_peoplerashes_people['DRIVER_ACTION'] == 'UNKNOWN') & (traffic_crashes_people['CELL_PHONE_USE'] == 'Y')].shape[0]

47

In [676]:
traffic_crashes_people[(traffic_crashes_people['DRIVER_ACTION'] == 'NONE') & (traffic_crashes_peoplerashes_people['CELL_PHONE_USE'] == 'Y')].shape[0]

55

In [677]:
traffic_crashes_people[(traffic_crashes_peoplerashes_people['DRIVER_ACTION'].isnull()) & (traffic_crashes_peoplerashes_people['CELL_PHONE_USE'] == 'Y')].shape[0]

0

we could find some rows that in CELL_PHONE_USE is yes , so we can change when DRIVER_ACTION is NONE or UNKOWN to CELL PHONE USE as it is very important cause of accidents , but we couldn't find any nan values in DRIVER_ACTION matches the value (yes) in CELL_PHONE_USE column. 

In [678]:
traffic_crashes_people.loc[
    (traffic_crashes_people['CELL_PHONE_USE'] == 'Y') & (traffic_crashes_people['DRIVER_ACTION'] == 'UNKNOWN'),
    'DRIVER_ACTION'] = 'CELL PHONE USE OTHER THAN TEXTING'

traffic_crashes_people.loc[
    (traffic_crashes_people['CELL_PHONE_USE'] == 'Y') & (traffic_crashes_people['DRIVER_ACTION'] == 'NONE'),
    'DRIVER_ACTION'] = 'CELL PHONE USE OTHER THAN TEXTING'

In [679]:
traffic_crashes_people['DRIVER_ACTION'].isnull().sum()

574

In [680]:
traffic_crashes_people['DRIVER_ACTION'].fillna('UNKNOWN',inplace=True)

In [681]:
traffic_crashes_people['DRIVER_VISION'].value_counts()

NOT OBSCURED              727187
UNKNOWN                   676534
OTHER                      14308
MOVING VEHICLES             8157
PARKED VEHICLES             5019
WINDSHIELD (WATER/ICE)      3966
BLINDED - SUNLIGHT          1704
TREES, PLANTS                576
BUILDINGS                    515
BLINDED - HEADLIGHTS         154
BLOWING MATERIALS            102
HILLCREST                     95
EMBANKMENT                    83
SIGNBOARD                     35
Name: DRIVER_VISION, dtype: int64

In [682]:
traffic_crashes_people.loc[
    (traffic_crashes_people['PERSON_TYPE'].isin(['PASSENGER','PEDESTRIAN','BICYCLE'])) & (traffic_crashes_people['DRIVER_VISION'].isnull()),
    'DRIVER_VISION'] = 'NOT RELEVANT'

In [683]:
traffic_crashes_people['DRIVER_VISION'].isnull().sum()

641

In [684]:
traffic_crashes_people['DRIVER_VISION'].fillna('UNKNOWN',inplace=True)

In [685]:
traffic_crashes_people['PHYSICAL_CONDITION'].isnull().sum()

367795

In [686]:
traffic_crashes_people['PHYSICAL_CONDITION'].value_counts()

NORMAL                          937000
UNKNOWN                         476369
IMPAIRED - ALCOHOL                6181
REMOVED BY EMS                    5280
OTHER                             4186
FATIGUED/ASLEEP                   3781
EMOTIONAL                         3547
ILLNESS/FAINTED                   1313
HAD BEEN DRINKING                 1050
IMPAIRED - DRUGS                   695
IMPAIRED - ALCOHOL AND DRUGS       384
MEDICATED                          178
Name: PHYSICAL_CONDITION, dtype: int64

###### We noticed that i have some conditions represnting if the person what alcoholic and drunk , and as we have another column with name BAC_RESULT VALUE we can extract some drunk people from their test value which says the person is drunk with value higher rthan 0.05.

In [687]:
traffic_crashes_people.loc[(traffic_crashes_people['PHYSICAL_CONDITION'] == "UNKNOWN") & (traffic_crashes_people['BAC_RESULT VALUE'] >= 0.05),'PHYSICAL_CONDITION'] = 'IMPAIRED - ALCOHOL'

In [688]:
traffic_crashes_people.loc[(traffic_crashes_people['PHYSICAL_CONDITION'] == "NORMAL") & (traffic_crashes_people['BAC_RESULT VALUE'] >= 0.05),'PHYSICAL_CONDITION'] = 'IMPAIRED - ALCOHOL'

In [689]:
traffic_crashes_people.loc[(traffic_crashes_people['PHYSICAL_CONDITION'].isnull()) & (traffic_crashes_people['BAC_RESULT VALUE'] >= 0.05),'PHYSICAL_CONDITION'] = 'IMPAIRED - ALCOHOL'

In [690]:
traffic_crashes_people['PHYSICAL_CONDITION'].isnull().sum()

367795

In [691]:
traffic_crashes_people['PHYSICAL_CONDITION'].fillna('UNKNOWN',inplace=True)

In [692]:
traffic_crashes_people['PEDPEDAL_ACTION'].isnull().sum()

1773341

In [693]:
traffic_crashes_people['PEDPEDAL_ACTION'].value_counts()

CROSSING - WITH SIGNAL                               7025
WITH TRAFFIC                                         5238
UNKNOWN/NA                                           4375
OTHER ACTION                                         4002
CROSSING - NO CONTROLS (NOT AT INTERSECTION)         2260
CROSSING - NO CONTROLS (AT INTERSECTION)             1900
NO ACTION                                            1875
CROSSING - AGAINST SIGNAL                            1643
CROSSING - CONTROLS PRESENT (NOT AT INTERSECTION)    1112
NOT AT INTERSECTION                                  1092
AGAINST TRAFFIC                                      1021
STANDING IN ROADWAY                                   800
TURNING LEFT                                          498
PARKED VEHICLE                                        406
ENTER FROM DRIVE/ALLEY                                330
WORKING IN ROADWAY                                    232
TURNING RIGHT                                         220
INTOXICATED PE

In [694]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'].isin(['DRIVER','NON-CONTACT VEHICLE'])) & (traffic_crashes_people['PEDPEDAL_ACTION'].isnull())].shape[0]

1406814

These columns PEDPEDAL_ACTION and the column PEDPEDAL_VISIBILITY and PEDPEDAL_LOCATION are special attribute when the person is pedestrian and bicyclist so let's add a new unique value for the drivers and passengers as they NOT RELEVANT for these columns.

In [695]:
traffic_crashes_people.loc[
    (traffic_crashes_people['PERSON_TYPE'].isin(['PASSENGER','DRIVER','NON-CONTACT VEHICLE'])) & (traffic_crashes_people['PEDPEDAL_ACTION'].isnull()),
    'PEDPEDAL_ACTION'] = 'NOT RELEVANT'

In [696]:
traffic_crashes_people['PEDPEDAL_ACTION'].fillna('UNKNOWN/NA',inplace=True)

In [697]:
traffic_crashes_people['PEDPEDAL_VISIBILITY'].isnull().sum()

1773404

In [698]:
traffic_crashes_people['PEDPEDAL_VISIBILITY'].value_counts()

NO CONTRASTING CLOTHING    27031
CONTRASTING CLOTHING        4471
OTHER LIGHT SOURCE USED     1977
REFLECTIVE MATERIAL          876
Name: PEDPEDAL_VISIBILITY, dtype: int64

In [699]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'].isin(['DRIVER','NON-CONTACT VEHICLE','PASSENGER'])) & (traffic_crashes_people['PEDPEDAL_VISIBILITY'].isnull())].shape[0]

1772389

In [700]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'].isin(['PEDESTRIAN','NON-MOTOR VEHICLE','BICYCLE'])) & (traffic_crashes_people['PEDPEDAL_VISIBILITY'].isnull())].shape[0]

1015

In [701]:
traffic_crashes_people.loc[
    (traffic_crashes_people['PERSON_TYPE'].isin(['PASSENGER','DRIVER','NON-CONTACT VEHICLE'])) & (traffic_crashes_people['PEDPEDAL_VISIBILITY'].isnull()),
    'PEDPEDAL_VISIBILITY'] = 'NOT RELEVANT'

In [702]:
traffic_crashes_people['PEDPEDAL_VISIBILITY'].fillna('UNKNOWN/NA',inplace=True)

In [703]:
traffic_crashes_people['PEDPEDAL_LOCATION'].value_counts()

IN ROADWAY         15131
IN CROSSWALK       11571
UNKNOWN/NA          2873
NOT IN ROADWAY      1527
BIKEWAY             1468
BIKE LANE           1153
DRIVEWAY ACCESS      490
SHOULDER             199
Name: PEDPEDAL_LOCATION, dtype: int64

In [704]:
traffic_crashes_people[(traffic_crashes_people['PERSON_TYPE'].isin(['DRIVER','NON-CONTACT VEHICLE','PASSENGER'])) & (traffic_crashes_people['PEDPEDAL_LOCATION'].isnull())].shape[0]

1772383

In [705]:
traffic_crashes_people.loc[
    (traffic_crashes_people['PERSON_TYPE'].isin(['PASSENGER','DRIVER','NON-CONTACT VEHICLE'])) & (traffic_crashes_people['PEDPEDAL_LOCATION'].isnull()),
    'PEDPEDAL_LOCATION'] = 'NOT RELEVANT'

In [706]:
traffic_crashes_people['PEDPEDAL_LOCATION'].isnull().sum()

964

In [707]:
traffic_crashes_people['PEDPEDAL_LOCATION'].fillna('UNKNOWN/NA',inplace=True)

In [708]:
traffic_crashes_people['BAC_RESULT'].value_counts()

TEST NOT OFFERED                   1418900
TEST REFUSED                         15005
TEST PERFORMED, RESULTS UNKNOWN       3481
TEST TAKEN                            2564
Name: BAC_RESULT, dtype: int64

In [709]:
traffic_crashes_people['BAC_RESULT'].isnull().sum()

367809

In [710]:
traffic_crashes_people['BAC_RESULT VALUE'].isnull().sum()

1805739

In [711]:
traffic_crashes_people['BAC_RESULT VALUE'].unique()

array([ nan, 0.21, 0.14, 0.2 , 0.12, 0.18, 0.19, 0.17, 0.24, 0.13, 0.58,
       0.1 , 0.  , 0.23, 0.03, 0.22, 0.08, 0.32, 0.15, 0.09, 0.25, 0.04,
       0.34, 0.11, 0.16, 0.38, 0.44, 0.3 , 0.33, 0.28, 0.27, 0.31, 0.05,
       0.06, 1.  , 0.26, 0.07, 0.41, 0.29, 0.02, 0.67, 0.36, 0.79, 0.35,
       0.88, 0.85, 0.37, 0.39, 0.01, 0.8 , 0.95, 0.6 , 0.45, 0.99, 0.47,
       0.4 , 0.98])

In [712]:
traffic_crashes_people[(traffic_crashes_people['BAC_RESULT'].isnull()) & (traffic_crashes_people['BAC_RESULT VALUE'].notnull())]

Empty DataFrame
Columns: [PERSON_ID, PERSON_TYPE, CRASH_RECORD_ID, VEHICLE_ID, CRASH_DATE, SEAT_NO, CITY, STATE, SEX, AGE, DRIVERS_LICENSE_STATE, DRIVERS_LICENSE_CLASS, SAFETY_EQUIPMENT, AIRBAG_DEPLOYED, EJECTION, INJURY_CLASSIFICATION, DRIVER_ACTION, DRIVER_VISION, PHYSICAL_CONDITION, PEDPEDAL_ACTION, PEDPEDAL_VISIBILITY, PEDPEDAL_LOCATION, BAC_RESULT, BAC_RESULT VALUE, CELL_PHONE_USE]
Index: []

No existing value for 'BAC_RESULT VALUE' when  the 'BAC_RESULT' is null. 

In [713]:
traffic_crashes_people['PHYSICAL_CONDITION'].unique()

array(['UNKNOWN', 'NORMAL', 'IMPAIRED - DRUGS', 'REMOVED BY EMS',
       'IMPAIRED - ALCOHOL', 'FATIGUED/ASLEEP', 'EMOTIONAL', 'OTHER',
       'HAD BEEN DRINKING', 'IMPAIRED - ALCOHOL AND DRUGS',
       'ILLNESS/FAINTED', 'MEDICATED'], dtype=object)

In [714]:
traffic_crashes_people[traffic_crashes_people['PHYSICAL_CONDITION'].isin(['IMPAIRED - ALCOHOL','HAD BEEN DRINKING','IMPAIRED - ALCOHOL AND DRUGS'])]['BAC_RESULT VALUE'].isnull().sum()

5858

OK, we can seta velue for exmaple like previous 0.05 when PHYSICAL_CONDITION is in (IMPAIRED - ALCOHOL,HAD BEEN DRINKING,IMPAIRED - ALCOHOL AND DRUGS) for the person which means he was drunk, but still about 5000 values so there are alot and alot of missing values yet, so we can drop this column. 

In [715]:
traffic_crashes_people = traffic_crashes_people.drop('BAC_RESULT VALUE', axis=1)

In [716]:
traffic_crashes_people['BAC_RESULT'].fillna('TEST NOT OFFERED',inplace=True)

We  extract what we can from the column CELL_PHONE_USE to assign some values to DRIVER_ACTION column , but no evidence to fill this column from any other columns except DRIVER_ACTION which it will just give me about 100 values not more. so we must drop this column also 

In [717]:
traffic_crashes_people = traffic_crashes_people.drop('CELL_PHONE_USE', axis=1)

In [718]:
info_df = pd.DataFrame()
info_df['missing_val'] = traffic_crashes_people.isnull().sum()
info_df['missing_val_ratio'] = (info_df['missing_val'] / traffic_crashes_people.shape[0] * 100).round().astype(int)
info_df

missing_val  missing_val_ratio
PERSON_ID                        0                  0
PERSON_TYPE                      0                  0
CRASH_RECORD_ID                  0                  0
VEHICLE_ID                       0                  0
CRASH_DATE                       0                  0
SEAT_NO                          0                  0
CITY                             0                  0
STATE                            0                  0
SEX                              0                  0
AGE                              0                  0
DRIVERS_LICENSE_STATE            0                  0
DRIVERS_LICENSE_CLASS            0                  0
SAFETY_EQUIPMENT                 0                  0
AIRBAG_DEPLOYED                  0                  0
EJECTION                         0                  0
INJURY_CLASSIFICATION            0                  0
DRIVER_ACTION                    0                  0
DRIVER_VISION                    0                  0
PHYSICAL_CONDITION               0                  0
PEDPEDAL_ACTION                  0                  0
PEDPEDAL_VISIBILITY              0                  0
PEDPEDAL_LOCATION                0                  0
BAC_RESULT                       0                  0

### Now we have our data with no missing values ..